Clarify top-level links

ethanresnick opened the pull request:
Addresses #250.

ethanresnick committed:
Clarify top-level links

Addresses #250

ethanresnick commented on format/index.md:
```diff
@@ -348,8 +348,9 @@ the referenced objects as a collection of resource objects.
-A top-level `"links"` object **MAY** be used to specify URL templates that can
-be used to formulate URLs for resources according to their type.
+A top-level `"links"` object **MAY** be used to specify URL templates that can be
+used to formulate URLs for resources. This often allows API responses to be more 
```
Here, I removed the phrase "according to their type" because it's incorrect—the templates are according to path—and added the next sentence to state the impetus for templates upfront.

dgeb replied:
How about "resource representations" instead of "API responses"?

ethanresnick replied:
Done!

ethanresnick commented on format/index.md:
```diff
@@ -371,14 +372,31 @@ For example:
+point of view of the reference document) that points at a repeated relationship. 
+Paths start with a particular resource type and can traverse related resources. For
+example `"posts.comments"` points at the `"comments"` relationship in each
+resource of type `"posts"`.
```
The above paragraph is what used to be [here](https://github.com/json-api/json-api/pull/262/files#diff-26e38355b69056b0db1c2b1c612241b2L403). Just moving it, and the subsequent paragraph, up, in order to state the behavior earlier in the section.

ethanresnick commented on format/index.md:
```diff
@@ -371,14 +372,31 @@ For example:
+However, a value in the `"links"` object **MAY** also be an object with two properties: 
+`"href"`and `"type"`. In this case, the `"href"` key holds the URL template while the 
+`"type"` key specifies the type of the linked resource(s), which facilitates lookups of 
+linked resource objects by the client.
```
Above paragraph is replacing the bullet point that was [here](https://github.com/json-api/json-api/pull/262/files#diff-26e38355b69056b0db1c2b1c612241b2L407) and gets at the crux of #250.

dgeb replied:
This phrasing seems a bit off: there is no concept of a "default". Each value may be either a string or an object. Similarly, there is no requirement that an object contains both `"href"` and `"type"` - it can contain either or both.

dgeb replied:
I think of a top level link object as a [collection object](http://jsonapi.org/format/#document-structure-collection-objects) that can't contain `"ids"`.

ethanresnick commented on format/index.md:
```diff
@@ -371,14 +372,31 @@ For example:
+    "posts.comments": {
+      "href": "http://example.com/comments/{posts.comments}",
+      "type": "comments"
+    }
```
Changed this so we have an example of both allowed formats in this section.

dgeb replied:
:+1:

ethanresnick commented on format/index.md:
```diff
@@ -398,17 +416,6 @@ the default explosion is to percent encode the array members (e.g. via
-  example `"posts.comments"` points at the `"comments"` relationship in each
-  resource of type `"posts"`.
-* The value of each key is interpreted as a URL template.
-* For each resource that the path points to, act as if it specified a
```
This bullet point was always pretty unclear. I tried to clean it up a bit at [line 454](https://github.com/json-api/json-api/pull/262/files#diff-26e38355b69056b0db1c2b1c612241b2R454) but it could still use to be more specific.

ethanresnick commented on format/index.md:
```diff
@@ -457,10 +469,6 @@ In a compound document, linked resources **MUST** be included as resource
 
-The type of each relationship **MAY** be specified in a resource-level or top-
-level `"links"` object with the `"type"` key. This facilitates lookups of linked
-resource objects by the client.
```
Removed this paragraph because it was now redundant: we said above that the top-level `"links"` can contain `"type"` keys, and the values at resource-level `"links"` are just resource or collection objects, which also allow `"type"` keys.

dgeb commented on format/index.md:
```diff
@@ -444,6 +451,11 @@ In this example, the URL for the author of all three posts is
+resource that the path points to, act as if that resource specified a relationship whose
+liked resource(s) can be found at the URL formed by expanding the URL template with values 
+drawn from the resource.
+
```
I think you're talking about expression expansion within URL templates here - is that right? Your reference to `key/path` is confusing. I'm not thrilled with the original phrasing in the bullet points, but I think this needs some work too.

ethanresnick replied:
I went through a few iterations of this and finally came up with something I'm relatively happy with. It's now at [line 390](https://github.com/json-api/json-api/pull/262/files#diff-26e38355b69056b0db1c2b1c612241b2R390)

ethanresnick committed:
Rewrite per feedback

ethanresnick committed:
Replace algorithmic explanation of applying the "links" object

Trying to give an algorithm was out of the character of the spec, which doesn't give precise algorithms anywhere else. And it just made the description super messy (without even being fully specific enough to really implement directly as written.)

ethanresnick committed:
fix typo

ethanresnick commented:
Hey @dgeb,

Sorry it took me a bit to get back to this. I've updated my PR based on your feedback.

dgeb commented on format/index.md:
```diff
@@ -371,14 +372,43 @@ For example:
+
+Each value in the `"links"` object may either be a string or an object. If a value is
+a string, that string represents a URL template. If the value is an object, it must have
+either an `"href"` key, a `"type"`key, or both; the `"href"` key holds a URL template
```
needs a space to break this up: `"type"`key

ethanresnick replied:
fixed!

dgeb commented:
@ethanresnick thanks for updating this. I'm now :+1: on these changes, which really explain the top-level `"links"` object more thoroughly than before. I only commented on one tiny typo in my read-through.

@steveklabnik @wycats what do you guys think?

ethanresnick committed:
fix typo

ethanresnick commented:
@dgeb Great!

dgeb commented:
Related to #276 - we may want to reach a consensus on `links` in general before clarifying the language.

dgeb commented:
Thanks for your efforts here, @ethanresnick. Ultimately we ended up simplifying top-level links a great deal in #341.
